schema-layer/data-structures/generics.md

```diff
@@ -0,0 +1,73 @@
+# Specification: IPLD Composites
+
+**Status: Prescriptive - Exploratory**
+
+Organizing IPLD data into usable, efficient, complex data structures spanning many blocks aimed for use by end-user applications.
+
+This document will re-use some terms found in the [IPLD data model](data-model-layer/data-model.md). 
+
+IPLD Composites offer codec agnostic programming interfaces for all common operations users can currently accomplish on Data-Model [“Kinds”](data-model-layer/data-model.md#kinds).
+
+Contents:
+
+  * [Motivation](#motivation)
+  * [Requirements](#requirements)
+  * [Composite Definition](#composite-definition)
+  * [Operations](#operations)
+
+## Motivation
+
+Even before the IPLD Data-Model was formally specified, developers were creating multi-block data-structures with similar semantics to single-block primitives. The most illustrative example of this is the `dag-pb` HAMT implementation used by IPFS for large directories.
+
+These early implementations of multi-block data structures exposed several problems.
+
+  * They are lacking self-description. A consumer of a graph containing these structures needs to have logic on top of IPLD and vary the way it performs operations on these data structures.
+  * Implementations of these data structures cannot perform operations on each other. In other words, multi-block data structures have a hard time building on top of each other.
+
+Since there wasn’t a standardized way to describe these data structures we couldn’t build libraries for paths and selectors that seamlessly supported them.
+
+As we started designing this system several other requirements surfaced.
+
+* Transparent encryption envelopes on the read **and** write.
+* Advanced `Link` types that can support some form of mutability and link to paths within other data structures.
+* Flexible multi-block binary types.
+
+## Requirements
+
+IPLD Composites cannot be implemented without:
+
+  * The IPLD Data-Model. While composites are codec agnostic they do require the full data model be implemented by the codec.
+
+## Composite Definition
+
+Composite definitions describe how to find an implementation of the data structure. When encoded into the data these also serve as a the self-description mechanism.
+
+***Open Issue: Fat Pointers***
+
+Early experiments simply reserved the `_type` property for composites to describe themselves. Reserving this property by default across any data in any block is highly problematic and makes it impossible to express certain data in IPLD.
+
+What we need in order to move forward is some extension/modification to `CID` in order to signal that “the data being linked to is a composite definition” at which point we can safely ad semantics to `_type` or other properties without reserving any property universally.
```

Comment:
The preceding paragraph is good, but the conclusion in this paragraph is not the only one possible from that setup, but phrases itself as if it is ('need').  I believe there are other solutions to this issue that aren't explored here and may be better.

Reply:
I think “need” is appropriate in the context of “fat pointers.” Unless there’s an alternative I’m not aware of, something in the CID needs to signal the use of a fat pointer, whether that’s a difference in codec, codec prefix, or CID version. If there’s another path to fat pointers that can be encoded into the data I’d love to hear it and incorporate it, but if this is just another branch of the “out of band signal” thread then I’m inclined to leave the statement as-is.

Reply:
does merging this document imply that were all in agreement that we will be using _type to convey type information?

Reply:
Now that I’m thinking about it, I need to add a paragraph in the “Definition” to talk about applying definitions, and start to map out how in-band and out-of-band signaling might work.

Reply:
> does merging this document imply that were all in agreement that we will be using _type to convey type information?

The document doesn’t actually say this. `_type` is defined as part of the “Composite Definition” which is being set aside so that we don’t reserve the property in a general way, it’s only defined as part of a “definition.” We’ve already conceded that there needs to be an additional signal to say “this is a definition” at which point we add these semantics to the property as to avoid reserving the property name.

Merging the document conveys that, in the experiments we’re doing, we’ll use a string `_type` (we already do) **in** definitions.

I updated the paragraph in the `_type` section to be a little more explicit about this only being “when a definition is applied.”

Reply:
Something about this sentence still bothers me, and it's difficult to put a finger on it.  It's not the "_type" string; not super concerned with the string itself, now that it's clear we're out of the blue-boxing trap.

Something seems fishy because we have examples of systems that work today without "extension or modification of CID".  The current gateways use HAMTs and other shard systems for dirs and files; and they didn't need an extension or modification of CID.  Filecoin is planning to use several fancy layouts; and they don't seem to be discussing an extension or modification of CID.  If we can't describe and reason about whatever these systems are doing, we've got a fairly serious blind spot.

Maybe some extension or modification of CID will be wise.  If fat pointers count as that (... do they?  not clear to me if all variations of fat pointers are necessarily modifications of CID), maybe they'll be wise.  But stating it as "need" and without alternatives seems wrong to me, given that we can observe systems evolving *right now* which very much don't seem to "need" this.  And I'm very wary of making a committal statement like "CIDs need to change".  (Maybe that's not how this is supposed to taken, but that's the idea I keep getting when I re-read the current wording.)

---

ANYWAY, I'm still thumbs up on merging this file overall under the [exploration-reports heading](https://github.com/ipld/specs/pull/126#issuecomment-510013092) in its current state, mind.  I'm just really very conscious of this cluster of neurons firing for me saying we're still missing something, even if I can't articulate it particularly well yet.  (I think I have some ideas, but I had to descend *way* deep into the Turing Tarpit to get them, and it might take a bit to write up readably.)

Reply:
So, my current thinking is that we probably will need an extension to CID, but given how much my thinking has evolved on this recently I’m going to back out definitive sounding language from the statement.